### PR TITLE
Radiation Filter Functor: Remove Macros

### PR DESCRIPTION
**Pull request by psychocoderHPC**
Cleanup the radiation plugin.
  - remove all radiation extensions in the PIConGPU core routines
  - remove macro's `RAD_MARK_PARTICLE` and `RAD_ACTIVATE_GAMMA_FILTER`
    - `RAD_ACTIVATE_GAMMA_FILTER`
      substituted with a user defined functor (can be use any attribute of an particle)
    - `RAD_MARK_PARTICLE`
      can be described by #1866
  - call particle filter within the radiation plugin
  - update radiation param files
- **param file change**: rename `radiationFlag` to `radiationMask`

`MySimulation`
 - update `momentumPrev1` within `runOneStep()`

- [x] rebase against #1876
- [x] runtime bunch tests

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
# bunch example test (timestep 3000)
## dev
bunch_dev

## pull request
bunch_pull

## pull requst with filter (all particle with gamma>=3)
- note: all particle in this simulation have a gamma of 5, therefore the plot is equal to the previous
bunch_pull_gamma3

## pull requst with filter (all particle with gamma>=30)
- all radiation data are zero, because all particle are to slow

File: src/picongpu/include/simulation_defines/param/radiationConfig.param [hunk 1]
```diff
@@ -105,32 +108,59 @@ namespace picongpu
...
+                if(
+                    picongpu::gamma<float_X>(
+                        particle[ picongpu::momentum_ ],
+                        picongpu::traits::attribute::getMass(
+                            particle[ picongpu::weighting_ ],
+                            particle
+                        )
+                    ) >= radiationGamma
+                )
+                particle[ picongpu::radiationFlag_ ] = true;
```
Comment (PrometheusPi) [inline src/picongpu/include/simulation_defines/param/radiationConfig.param]: QUESTION:
How do you make sure it is initialized as `false`?
(I could not find the corresponding param file.)

File: src/picongpu/include/simulationControl/MySimulation.hpp [hunk 2]
```diff
@@ -480,6 +483,28 @@ class MySimulation : public SimulationHelper<simDim>
     {
         namespace nvfct = PMacc::nvidia::functors;
 
+        typedef typename PMacc::particles::traits::FilterByIdentifier
+        <
+            VectorAllSpecies,
+            momentumPrev1
+        >::type VectorSpeciesWithMementumPrev1;
+
+        /* copy attribute momentum to momentumPrev1 */
```
Comment (PrometheusPi) [inline src/picongpu/include/simulationControl/MySimulation.hpp]: OPTIONAL:
Please add a comment that this copy is **before** the pusher and that copy **and** pusher are **before** the radiation plugin.

Comment (psychocoderHPC) [inline src/picongpu/include/simulationControl/MySimulation.hpp]: IMO this is not the right place for this comment, the order of the plugin calls and the runOneStep should commented in `PMacc::SimulationHelper`.
The order inside run one step is visible by the calls itself.

File: src/picongpu/include/simulation_defines/param/radiationConfig.param (hunk 1 above)
Comment (psychocoderHPC) [inline src/picongpu/include/simulation_defines/param/radiationConfig.param]: The default behavior is defined in `speciesAttributes.pram`

File: src/picongpu/include/plugins/radiation/GetRadiationFlag.hpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    template< typename T_Particle >
+    HDINLINE bool getRadiationFlag( const T_Particle& particle )
+    {
+        constexpr bool hasRadiationFlag = PMacc::traits::HasFlag<
+            typename T_Particle::FrameType,
+            radiationFlag
+        >::type::value;
+        return GetRadiationFlag< hasRadiationFlag >{}( particle );
+    }
+} // namespace picongpu
```
Comment (psychocoderHPC) [inline src/picongpu/include/plugins/radiation/GetRadiationFlag.hpp]: is not required by our coding style^^, could I skip this? The begin of the namespace also does not contain a empty line.

File: src/picongpu/include/plugins/radiation/ExecuteParticleFilter.hpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+         *
+         * @param particle particle to be used
+         * @return value of the attribute `radiationFlag`
+         */
+        template< typename T_Species >
+        void operator()( std::shared_ptr<T_Species> species, const uint32_t currentStep )
+        {
+            auto radiationEnableFunctor = picongpu::radiation::FilterEnableRadiation( currentStep );
+            species->manipulateAllParticles(
+                currentStep,
```
Comment (psychocoderHPC) [inline src/picongpu/include/plugins/radiation/ExecuteParticleFilter.hpp]: it is allowed to add a constructor to the filter functor. 
```
Filter(const unit32_t currentStep);
```
This allows to write time dependent filter.

**Comment by ax3l** [comment]
@PrometheusPi can you please update your docs regarding `radiationFlag` -> `radiationMask`?

File: examples/Bunch/include/simulation_defines/param/radiationConfig.param [hunk 5]
```diff
@@ -112,30 +115,45 @@ namespace parameters
...
+            if(
+                picongpu::gamma<float_X>(
+                    particle[ picongpu::momentum_ ],
+                    picongpu::traits::attribute::getMass(
+                        particle[ picongpu::weighting_ ],
+                        particle
+                    )
+                ) >= radiationGamma
+            )
+            particle[ picongpu::radiationMask_ ] = true;
```
Comment (PrometheusPi) [inline examples/Bunch/include/simulation_defines/param/radiationConfig.param]: REQUIRED:
This is still misaligned

**Comment by PrometheusPi** [comment]
Wiki is adjusted.

File: examples/Bunch/include/simulation_defines/param/radiationConfig.param (hunk 5 above)
Comment (psychocoderHPC) [inline examples/Bunch/include/simulation_defines/param/radiationConfig.param]: true

File: src/picongpu/include/simulationControl/MySimulation.hpp [hunk 6]
```diff
@@ -480,6 +483,28 @@ class MySimulation : public SimulationHelper<simDim>
...
+            particles::Manipulate<
+                particles::manipulators::CopyAttribute<
+                    momentumPrev1,
+                    momentum
+                >,
+                bmpl::_1
+            >
+        > copyMomentumPrev1;
+        copyMomentumPrev1(
+            forward( particleStorage ),
```
Comment (ax3l) [inline src/picongpu/include/simulationControl/MySimulation.hpp]: @psychocoderHPC careful, this class is removed in `dev` now. access species via `DataConnector` :)

**Comment by psychocoderHPC** [comment]
fixed compile issu

**Pull request merged**